mesonbuild/build.py

```diff
@@ -1086,10 +1086,17 @@ def is_internal(self):
 
     def link(self, target):
         for t in listify(target, unholder=True):
-            if isinstance(self, StaticLibrary) and self.need_install and t.is_internal():
-                # When we're a static library and we link_with to an
-                # internal/convenience library, promote to link_whole.
-                return self.link_whole(t)
+            if isinstance(self, StaticLibrary) and self.need_install:
+                if isinstance(t, (CustomTarget, CustomTargetIndex)):
+                    if (isinstance(t, CustomTarget) and not t.should_install()) or (isinstance(t, CustomTargetIndex) and not t.target.should_install()):
```

Comment:
I would implement `is_internal()` and `should_install()` on CustomTarget and CustomTargetIndex, and use `t.should_install()` instead of `t.need_install` here. `is_internal()` on custom target should check if its filename ends with '.a' which the only clue you have to know it's actually a static library. If you do that, you can still let it call self.link_while(t) here, because that case is already handled in link_whole() by raising an exception.

Reply:
I don't know about meson internals, but on windows for example one could have a custom_target producing a static library that does not end with `.a`, since `.lib` is standard on that platform.

Reply:
I think this is fine as-is. I would change `CustomTargetIndex` and add a `should_install()` method to it, then the code here can be simplified to:

```py
if isinstance(t, (CustomTarget, CustomTargetIndex)):
    if not t.should_install():
        mlog.warning(...)
    self.link_targets.append(t)
elif t.is_internal():
```

And it will keep working if/when we add support for only installing some outputs of a custom target.